include/treelite/logging.h

```diff
@@ -9,9 +9,144 @@
 
 #include <treelite/thread_local.h>
 #include <iostream>
+#include <stdexcept>
+#include <string>
+#include <sstream>
+#include <memory>
+#include <cstdio>
+#include <ctime>
 
 namespace treelite {
 
+/*!
+ * \brief Exception class that will be thrown by Treelite
+ */
+struct Error : public std::runtime_error {
+  explicit Error(const std::string& s) : std::runtime_error(s) {}
+};
+
+template <typename X, typename Y>
+std::unique_ptr<std::string> LogCheckFormat(const X& x, const Y& y) {
+  std::ostringstream os;
+  os << " (" << x << " vs. " << y << ") ";
+  /* CHECK_XX(x, y) requires x and y can be serialized to string. Use CHECK(x OP y) otherwise. */
+  return std::make_unique<std::string>(os.str());
+}
+
+#if defined(__GNUC__) || defined(__clang__)
+#define TREELITE_ALWAYS_INLINE inline __attribute__((__always_inline__))
+#elif defined(_MSC_VER)
+#define TREELITE_ALWAYS_INLINE __forceinline
+#else
+#define TREELITE_ALWAYS_INLINE inline
+#endif
+
+#define DEFINE_CHECK_FUNC(name, op)                                                        \
+  template <typename X, typename Y>                                                        \
+  TREELITE_ALWAYS_INLINE std::unique_ptr<std::string> LogCheck##name(const X& x, const Y& y) { \
+    if (x op y) return nullptr;                                                            \
+    return LogCheckFormat(x, y);                                                           \
+  }                                                                                        \
+  TREELITE_ALWAYS_INLINE std::unique_ptr<std::string> LogCheck##name(int x, int y) {           \
+    return LogCheck##name<int, int>(x, y);                                                 \
+  }
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsign-compare"
+DEFINE_CHECK_FUNC(_LT, <)
+DEFINE_CHECK_FUNC(_GT, >)
+DEFINE_CHECK_FUNC(_LE, <=)
+DEFINE_CHECK_FUNC(_GE, >=)
+DEFINE_CHECK_FUNC(_EQ, ==)
+DEFINE_CHECK_FUNC(_NE, !=)
+#pragma GCC diagnostic pop
+
+
+#define CHECK_BINARY_OP(name, op, x, y)                  \
+  if (auto __treelite__log__err = treelite::LogCheck##name(x, y))  \
+      treelite::LogMessageFatal(__FILE__, __LINE__).stream() \
```

Comment:
Might a a good idea to use `::treelite:: ...` to specify we should start with global namespace.

Reply:
Done